### PR TITLE
Add thrift transport implementation (framed + unframed)

### PR DESCRIPTION
**Pull request by phile314**
This adds a dependency on the network package, but that package is almost always installed anyways.

### CODE REVIEW

File: src/Pinch/Transport.hs [hunk 1]
```diff
@@ -0,0 +1,124 @@
+{-# LANGUAGE RankNTypes #-}
+
+module Pinch.Transport
+  ( Transport(..)
+  , framedTransport
+  , unframedTransport
+  , Connection(..)
+  , ReadResult(..)
+  )where
```
Comment (abhinav) [inline src/Pinch/Transport.hs:9]:
```suggestion
  ) where
```

File: src/Pinch/Transport.hs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+  deriving (Eq, Show)
+
+-- | A bidirectional transport to read/write messages from/to.
+data Transport
+  = Transport
+  { writeMessage :: B.Builder -> IO ()
+  , readMessage  :: forall a . G.Get a -> IO (ReadResult a)
+  }
+
+-- | Creates a thrift framed transport. See also https://github.com/apache/thrift/blob/master/doc/specs/thrift-rpc.md#framed-vs-unframed-transport .
```
Comment (abhinav) [inline src/Pinch/Transport.hs:48]:
```suggestion
-- | Creates a thrift framed transport. See also https://github.com/apache/thrift/blob/master/doc/specs/thrift-rpc.md#framed-vs-unframed-transport.
```

File: src/Pinch/Transport.hs [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        let sz = fromIntegral <$> G.runGet G.getInt32be szBs
+        case sz of
+          Right x -> do
+            msgBs <- getExactly c x
+            if BS.length msgBs < x
+              then
+                -- less data has been returned than expected. This means we have reached EOF.
+                pure $ RREOF
+              else
+                pure $ either RRFailure RRSuccess $ G.runGet p msgBs
```
Comment (abhinav) [inline src/Pinch/Transport.hs:70]: optionally, we can put `pure` before the `if`:
```suggestion
            pure $ if BS.length msgBs < x
              then
                -- less data has been returned than expected. This means we have reached EOF.
                RREOF
              else
                either RRFailure RRSuccess $ G.runGet p msgBs
```

File: src/Pinch/Transport.hs [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+          Left s -> pure $ RRFailure $ "Invalid frame size: " ++ show s
+
+-- | Creates a thrift unframed transport. See also https://github.com/apache/thrift/blob/master/doc/specs/thrift-rpc.md#framed-vs-unframed-transport .
```
Comment (abhinav) [inline src/Pinch/Transport.hs:73]:
```suggestion
-- | Creates a thrift unframed transport. See also https://github.com/apache/thrift/blob/master/doc/specs/thrift-rpc.md#framed-vs-unframed-transport.
```

**Review by abhinav — APPROVED** [review]
Minor stylistic suggestions but LGTM otherwise.

**Comment by phile314** [comment]
I updated the PR with all your suggestions.

**Review by abhinav — APPROVED** [review]

**Pull request merged**